File: sample/cpp/07_game_of_life_loop.cpp
#include "cpp_module/py_runtime.h"




list<list<int64>> next_state(list<list<int64>> grid, int64 w, int64 h) {
    list<int64> row;
    int64 cnt;
    int64 nx;
    int64 ny;
    int64 alive;
    
    list<list<int64>> nxt = list<list<int64>>{};
    for (int64 y = 0; y < h; ++y) {
        row = list<int64>{};
        for (int64 x = 0; x < w; ++x) {
            cnt = 0;
            for (int64 dy = -1; dy < 2; ++dy) {
                for (int64 dx = -1; dx < 2; ++dx) {
                    if ((dx != 0) || (dy != 0)) {
                        nx = (x + dx + w) % w;
                        ny = (y + dy + h) % h;
                        cnt += grid[ny][nx];
                    }
                }
            }
            alive = grid[y][x];
            if ((alive == 1) && ((cnt == 2) || (cnt == 3))) {
                row.push_back(1);
            } else {
                if ((alive == 0) && (cnt == 3))
                    row.push_back(1);
                else
                    row.push_back(0);
            }
        }
        nxt.push_back(row);
    }
    return nxt;
}

list<uint8> render(list<list<int64>> grid, int64 w, int64 h, int64 cell) {
    int64 v;
    int64 base;
    
    int64 width = w * cell;
    int64 height = h * cell;
    list<uint8> frame = list<uint8>(width * height);
    for (int64 y = 0; y < h; ++y) {
        for (int64 x = 0; x < w; ++x) {
            v = (grid[y][x] ? 255 : 0);
            for (int64 yy = 0; yy < cell; ++yy) {
                base = (y * cell + yy) * width + x * cell;
                for (int64 xx = 0; xx < cell; ++xx)
                    frame[base + xx] = v;
            }
        }
    }
    return list<uint8>(frame);
}

void run_07_game_of_life_loop() {
    list<int64> row;
    int64 noise;
    int64 kind;
    int64 ph;
    int64 pw;
    list<list<int64>> grid;
    int64 _;
    int64 py;
    int64 px;
    
    int64 w = 144;
    int64 h = 108;
    int64 cell = 4;
    int64 steps = 210;
    str out_path = "sample/out/07_game_of_life_loop.gif";
    
    float64 start = perf_counter();
    grid = list<list<int64>>{};
    for (_ = 0; _ < h; ++_) {
        row = list<int64>{};
        for (_ = 0; _ < w; ++_)
            row.push_back(0);
        grid.push_back(row);
    }
    
    
    
    for (int64 y = 0; y < h; ++y) {
        for (int64 x = 0; x < w; ++x) {
            noise = (x * 37 + y * 73 + x * y % 19 + (x + y) % 11) % 97;
            if (noise < 3)
                grid[y][x] = 1;
        }
    }
    
    
    list<list<int64>> glider = list<list<int64>>{list<int64>{0, 1, 0}, list<int64>{0, 0, 1}, list<int64>{1, 1, 1}};
    list<list<int64>> r_pentomino = list<list<int64>>{list<int64>{0, 1, 1}, list<int64>{1, 1, 0}, list<int64>{0, 1, 0}};
    list<list<int64>> lwss = list<list<int64>>{list<int64>{0, 1, 1, 1, 1}, list<int64>{1, 0, 0, 0, 1}, list<int64>{0, 0, 0, 0, 1}, list<int64>{1, 0, 0, 1, 0}};
    
    for (int64 gy = 8; gy < h - 8; gy += 18) {
        for (int64 gx = 8; gx < w - 8; gx += 22) {
            kind = (gx * 7 + gy * 11) % 3;
            if (kind == 0) {
                ph = py_len(glider);
                for (py = 0; py < ph; ++py) {
                    pw = py_len(glider[py]);
                    for (px = 0; px < pw; ++px) {
                        if (glider[py][px] == 1)
                            grid[(gy + py) % h][(gx + px) % w] = 1;
                    }
                }
            } else {
                if (kind == 1) {
                    ph = py_len(r_pentomino);
                    for (py = 0; py < ph; ++py) {
                        pw = py_len(r_pentomino[py]);
                        for (px = 0; px < pw; ++px) {
                            if (r_pentomino[py][px] == 1)
                                grid[(gy + py) % h][(gx + px) % w] = 1;
                        }
                    }
                } else {
                    ph = py_len(lwss);
                    for (py = 0; py < ph; ++py) {
                        pw = py_len(lwss[py]);
                        for (px = 0; px < pw; ++px) {
                            if (lwss[py][px] == 1)
                                grid[(gy + py) % h][(gx + px) % w] = 1;
                        }
                    }
                }
            }
        }
    }
    
    list<list<uint8>> frames = list<list<uint8>>{};
    for (_ = 0; _ < steps; ++_) {
        frames.push_back(render(grid, w, h, cell));
        grid = next_state(grid, w, h);
    }
    
    save_gif(out_path, w * cell, h * cell, frames, grayscale_palette(), 4, 0);
    float64 elapsed = perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", steps);
    py_print("elapsed_sec:", elapsed);
}

int main() {
    run_07_game_of_life_loop();
    return 0;
}

File: sample/rs/07_game_of_life_loop.rs
mod py_runtime;
pub use crate::py_runtime::{math, pytra, time};
use crate::py_runtime::*;

use crate::time::perf_counter;
use crate::pytra::utils::gif::grayscale_palette;
use crate::pytra::utils::gif::save_gif;

// 07: Sample that outputs Game of Life evolution as a GIF.

fn next_state(grid: &[Vec<i64>], w: i64, h: i64) -> Vec<Vec<i64>> {
    let mut nxt: Vec<Vec<i64>> = vec![];
    for y in (0)..(h) {
            let mut row: Vec<i64> = vec![];
            for x in (0)..(w) {
                    let mut cnt = 0;
                    for dy in (-1)..(2) {
                            for dx in (-1)..(2) {
                                    if (dx != 0) || (dy != 0) {
                                        let nx = (x + dx + w) % w;
                                        let ny = (y + dy + h) % h;
                                        cnt += grid[((if ((ny) as i64) < 0 { (grid.len() as i64 + ((ny) as i64)) } else { ((ny) as i64) }) as usize)][((if ((nx) as i64) < 0 { (grid[((if ((ny) as i64) < 0 { (grid.len() as i64 + ((ny) as i64)) } else { ((ny) as i64) }) as usize)].len() as i64 + ((nx) as i64)) } else { ((nx) as i64) }) as usize)];
                                    }
                            }
                    }
                    let alive = grid[((y) as usize)][((x) as usize)];
                    if (alive == 1) && ((cnt == 2) || (cnt == 3)) {
                        row.push(1);
                    } else if (alive == 0) && (cnt == 3) {
                        row.push(1);
                    } else {
                        row.push(0);
                    }
            }
            nxt.push(row);
    }
    return nxt;
}

fn render(grid: &[Vec<i64>], w: i64, h: i64, cell: i64) -> Vec<u8> {
    let width = w * cell;
    let height = h * cell;
    let mut frame = vec![0u8; (width * height) as usize];
    for y in (0)..(h) {
            for x in (0)..(w) {
                    let v = (if grid[((y) as usize)][((x) as usize)] != 0 { 255 } else { 0 });
                    for yy in (0)..(cell) {
                            let base = (y * cell + yy) * width + x * cell;
                            for xx in (0)..(cell) {
                                    let __idx_i64_8 = ((base + xx) as i64);
                                    let __idx_7 = if __idx_i64_8 < 0 { (frame.len() as i64 + __idx_i64_8) as usize } else { __idx_i64_8 as usize };
                                    frame[__idx_7] = ((v) as u8);
                            }
                    }
            }
    }
    return frame;
}

fn run_07_game_of_life_loop() {
    let w = 144;
    let h = 108;
    let cell = 4;
    let steps = 105;
    let out_path = ("sample/out/07_game_of_life_loop.gif").to_string();
    
    let start = perf_counter();
    let mut grid: Vec<Vec<i64>> = (((0)..(h))).map(|py_underscore| vec![0; ((w) as usize)]).collect::<Vec<_>>();
    
    // Lay down sparse noise so the whole field is less likely to stabilize too early.
    // Avoid large integer literals so all transpilers handle the expression consistently.
    for y in (0)..(h) {
            for x in (0)..(w) {
                    let noise = (x * 37 + y * 73 + x * y % 19 + (x + y) % 11) % 97;
                    if noise < 3 {
                        let __idx_11 = ((y) as usize);
                        let __idx_12 = ((x) as usize);
                        grid[__idx_11][__idx_12] = 1;
                    }
            }
    }
    // Place multiple well-known long-lived patterns.
    let glider = vec![vec![0, 1, 0], vec![0, 0, 1], vec![1, 1, 1]];
    let r_pentomino = vec![vec![0, 1, 1], vec![1, 1, 0], vec![0, 1, 0]];
    let lwss = vec![vec![0, 1, 1, 1, 1], vec![1, 0, 0, 0, 1], vec![0, 0, 0, 0, 1], vec![1, 0, 0, 1, 0]];
    
    let mut gy: i64 = 8;
    while gy < h - 8 {
        let mut gx: i64 = 8;
        while gx < w - 8 {
            let kind = (gx * 7 + gy * 11) % 3;
            if kind == 0 {
                let mut ph = glider.len() as i64;
                for py in (0)..(ph) {
                        let mut pw = (glider[((py) as usize)]).clone().len() as i64;
                        for px in (0)..(pw) {
                                if glider[((py) as usize)][((px) as usize)] == 1 {
                                    let __idx_15 = (((gy + py) % h) as usize);
                                    let __idx_16 = (((gx + px) % w) as usize);
                                    grid[__idx_15][__idx_16] = 1;
                                }
                        }
                }
            } else if kind == 1 {
                let mut ph = r_pentomino.len() as i64;
                for py in (0)..(ph) {
                        let mut pw = (r_pentomino[((py) as usize)]).clone().len() as i64;
                        for px in (0)..(pw) {
                                if r_pentomino[((py) as usize)][((px) as usize)] == 1 {
                                    let __idx_19 = (((gy + py) % h) as usize);
                                    let __idx_20 = (((gx + px) % w) as usize);
                                    grid[__idx_19][__idx_20] = 1;
                                }
                        }
                }
            } else {
                let mut ph = lwss.len() as i64;
                for py in (0)..(ph) {
                        let mut pw = (lwss[((py) as usize)]).clone().len() as i64;
                        for px in (0)..(pw) {
                                if lwss[((py) as usize)][((px) as usize)] == 1 {
                                    let __idx_23 = (((gy + py) % h) as usize);
                                    let __idx_24 = (((gx + px) % w) as usize);
                                    grid[__idx_23][__idx_24] = 1;
                                }
                        }
                }
            }
            gx += 22;
        }
        gy += 18;
    }
    let mut frames: Vec<Vec<u8>> = vec![];
    for py_underscore in (0)..(steps) {
            frames.push(render(&(grid), w, h, cell));
            grid = next_state(&(grid), w, h);
    }
    save_gif(&(out_path), w * cell, h * cell, &(frames), &(grayscale_palette()), 4, 0);
    let elapsed = perf_counter() - start;
    println!("{} {}", ("output:").to_string(), out_path);
    println!("{} {}", ("frames:").to_string(), steps);
    println!("{} {}", ("elapsed_sec:").to_string(), elapsed);
}

fn main() {
    run_07_game_of_life_loop();
}
